Bug Squashes, Eastern Name Order, Gender Issues, and ItemTagResolver Display Tweaks

gattsuru opened the pull request:
Closes #6401 , some related PVP-mode issues.
Closes #6323
Closes #4587 , if confirmed that we don't want the texture update for Swineherd and Courier
Closes https://github.com/ldtteam/minecolonies-features/issues/244

# Changes proposed in this pull request:
- Moves logic related to the PVP-mode management of the TownHall block to BlockHutTownHall. Cleans up the sendPlayerMessages to insert arguments into TranslationTextComponents properly. Removes the PlayerEvent.BreakSpeed eventhandler, which only had logic related to the TownHall block but was getting called for the vast majority of blocks but not BlockHuts.
- Fixes double-counting of enchantment levels on fishing rods.
- Changes gendering for initial settlers spawned by the TownHall to spawn men if the town has more women, and to spawn women if the town has more men.  Also changes the first colonist to have a random chance of being male or female. Pattern will normally be female-male-male-female or male-female-male-female, but will recover from gender-seeking zombies a little more elegantly.
- Removes duplicate placeConstructionTape in addNewBuilding, which was causing it to fire for IRSComponentBlocks like the postbox/stash (and twice for normal buildings).
- Changes some model files to handle the 1.15.2+ Chest texture file change.  Warehouse, Shepherd, Chicken Farmer.  This could also be done for Swineherd and Courier, but they kinda look better as-is, since they're closed and the only effect is a shaded spot on the top of their chests.  Did fix the under-latch shadow textures for the Courier.
- Sets ItemTagResolvers to give the translated text for items that meet their requirements instead of the long Minecraft name of the tag (#com.minecraft.items.tags.Flowers).  Only impacts Beekeeper for now.  Not a perfect fix, as the StandardInteraction will show the first eligible item in the request (for vanilla flowers, dandelions), but solves the problem well enough without requiring deep changes.
- Adds a configuration option for Eastern Name Order.  For now, just switches to surname+lastname without space.  There are some Asian cultures that use more complicated structures and a whole mess of complications from social norms, but this'll handle the simple good-enough version.

Review please

gattsuru committed:
Bug Squashes, Eastern Name Order, and Gender Issues

Moves logic related to the PVP-mode management of the TownHall block to BlockHutTownHall. Cleans up the sendPlayerMessages to insert arguments into TranslationTextComponents properly. Removes the PlayerEvent.BreakSpeed, which only had logic related to the TownHall block but was getting called for everything except blockhuts.

Fixes double-counting of enchantment levels on fishing rods.

Changes gendering for initial settlers spawned by the TownHall to spawn men if the town has more women, and to spawn women if the town has more men.  Also changes the first colonist to have a random chance of being male or female.  Pattern will normally be female-male-male-female or male-female-male-female.

Removes duplicate placeConstructionTape in addNewBuilding, which was causing it to fire for  IRSComponentBlocks like the postbox and twice for normal buildings.

Changes some model files to handle the 1.15.2+ Chest texture file change.  Warehouse, Shepherd, Chicken Farmer.  This could also be done for Swineherd and Courier, but they kinda look better as-is, since they're closed and the only effect is a shaded spot on the top of their chests.  Did fix the under-latch shadow textures for the Courier.

Sets ItemTagResolvers to give the translated text for items that meet their requirements instead of the long Minecraft name of the tag (#com.minecraft.items.tags.Flowers).  Only impacts Beekeeper for now.

Adds a configuration option for Eastern Name Order.  For now, just switches to surname+lastname without space.  There are some Asian cultures that use more complicated structures and a whole mess of complications from social norms, but this'll handle the simple good-enough version.

gattsuru committed:
Improved ItemTagRequest handling

Improves handling of ItemTagRequest shortDisplayStrings and equivalent presentation in AbstractWindowRequestTree.

Removes an unnecessary space and unused import.

ravenbuilder934 commented on src/main/resources/assets/minecolonies/lang/en_us.json:
```diff
@@ -55,6 +55,9 @@
   "minecolonies.config.usemiddleinitial.comment": "Should citizens have a middle initial?",
 
+  "minecolonies.config.useeasternnameorder": "Eastern Name Order",
+  "minecolonies.config.useeasternnameorder.comment": "Should new citizens use Eastern name order, with surname first?",
```
```suggestion
  "minecolonies.config.useeasternnameorder.comment": "Should new citizens use the Eastern name order (with their surname first)?",
```

gattsuru committed:
Better description for useeasternnameorder

Improves description for eastern name order setting.

Co-authored-by: ravenbuilder934 <sabrinajamie@outlook.com>

Raycoms commented on src/main/java/com/minecolonies/coremod/colony/managers/CitizenManager.java:
```diff
@@ -543,10 +543,22 @@ public void onColonyTick(final IColony colony)
+                // 50-50 chance for first citizen to be male or female.
+                if(colony.getCitizenManager().getCitizens().isEmpty() && random.nextBoolean())
+                {
+                    femaleCount--;
```
that seems like a hacky solution, why not make this a bigger if/else and setIsFemale(true); else, the rest of the code

Raycoms commented on src/main/java/com/minecolonies/coremod/colony/managers/CitizenManager.java:
```diff
@@ -543,10 +543,22 @@ public void onColonyTick(final IColony colony)
-                if (citizens.size() % 2 == 0)
+                // Use double division to avoid getting two male colonists in a row for the first set.
+                // Have to subtract one from the citizen size, since trying to make comparison before including this colonist.
+                if (femaleCount < ((colony.getCitizenManager().getCitizens().size() - 1) / 2d))
```
2.0 instead of 2d, I'd say.

Raycoms commented on src/main/java/com/minecolonies/coremod/colony/requestsystem/requests/StandardRequests.java:
```diff
@@ -186,7 +186,16 @@ public ITextComponent getShortDisplayString()
-            combined.append(new StringTextComponent("#" + getRequest().getTag().toString()));
+            // getRequest().getTag() is a long string that can't be easily be read by players or turned into a translation key.
+            // Instead, preferentially select the first item from the list.
+            if(!stacks.isEmpty())
```
I'm wondering about this, I'd like the input of @OrionDevelopment on this, but all our tag requests are very specific. We could have something like com.minecolonies.coremod.tag.lastPartOfTagKey in the lang file and then use that, so we got it nicely translated

gattsuru replied:
That's definitely doable for internal use, and there are only a few item tags used at all right now.  Do you have any concerns about other mods or user/modpack-maker defined recipes or tags?

With the other changes in this commit, the getShortDisplayName version only shows up on the StandardInteraction message.  AbstractWindowRequestTree cycles through the items that will fulfill the request.  Do you prefer that, or would you rather show this new tag translated text?

Raycoms replied:
We could always do a check, since this is on the client side, we can check if we have a propper translation, if so we use it, else we use the other.

Raycoms commented on src/main/java/com/minecolonies/coremod/permissions/ColonyPermissionEventHandler.java:
```diff
@@ -214,7 +197,7 @@ public void on(final BlockEvent.BreakEvent event)
             }
 
-            if (event.getState().getBlock() == ModBlocks.blockHutTownHall && !validTownHallBreak && !event.getPlayer().isCreative())
+            if (event.getState().getBlock() == ModBlocks.blockHutTownHall && !((BlockHutTownHall)event.getState().getBlock()).getValidBreak() && !event.getPlayer().isCreative())
```
I definitely like this move, much better separation of concerns

gattsuru committed:
Code Cleanup on Gender Issues, TagRequest shortDisplayText

Cleans up the CitizenManager gendering code.

Changes ItemTagRequest.getShortDisplayName to use a TranslationText if available.  This will only show up on StandardInteractions for now.

Raycoms commented on src/main/java/com/minecolonies/coremod/colony/managers/CitizenManager.java:
```diff
@@ -542,11 +542,21 @@ public void onColonyTick(final IColony colony)
+                // For first citizen, give a random chance of male or female.
+                if (getCitizens().size() == 1)
+                {
+                    newCitizen.setIsFemale(random.nextBoolean());
```
Much better

Raycoms commented:
:shipit:

gattsuru committed:
Fixing duplicate json entry.

ravenbuilder934 committed:
Merge branch 'version/1.16.3' into version/1.16-fix-6401
